File: duckdata/crosslink/python/crosslink/flight_example.py
#!/usr/bin/env python3
"""
CrossLink Flight API Example

This script demonstrates how to use the Apache Arrow Flight API in CrossLink
for distributed data sharing between nodes.
"""

import os
import sys
import time
import argparse
import pandas as pd
import pyarrow as pa
from duckdata.crosslink import get_instance, CrossLinkConfig, OperationMode

def create_sample_table():
    """Create a sample table for demonstration."""
    df = pd.DataFrame({
        'id': [1, 2, 3, 4, 5],
        'name': ['Alice', 'Bob', 'Charlie', 'David', 'Eve'],
        'value': [10.1, 20.2, 30.3, 40.4, 50.5]
    })
    return pa.Table.from_pandas(df)

def main():
    parser = argparse.ArgumentParser(description="CrossLink Flight API Example")
    parser.add_argument("--mode", choices=['server', 'client'], required=True,
                        help="Run as server or client")
    parser.add_argument("--host", default="localhost",
                        help="Host for the Flight server (default: localhost)")
    parser.add_argument("--port", type=int, default=8815,
                        help="Port for the Flight server (default: 8815)")
    args = parser.parse_args()

    if args.mode == 'server':
        # Server mode: start a Flight server and share a dataset
        print(f"Starting server on {args.host}:{args.port}")
        
        # Create a configuration for distributed mode
        config = CrossLinkConfig(
            db_path="crosslink_server.duckdb",
            debug=True,
            mode=OperationMode.DISTRIBUTED,
            flight_host=args.host,
            flight_port=args.port
        )
        
        # Initialize CrossLink
        cl = get_instance(config=config)
        
        # Start the Flight server
        if cl.start_flight_server():
            print(f"Flight server started on port {cl.flight_server_port()}")
        else:
            print("Failed to start Flight server")
            return 1
            
        try:
            # Create and share a sample table
            table = create_sample_table()
            print("\nCreated sample table:")
            print(table.to_pandas())
            
            # Push the table to local storage
            dataset_id = cl.push(table, name="sample_dataset", 
                                description="Sample dataset for Flight demo")
            print(f"Table pushed with ID: {dataset_id}")
            
            # List available datasets
            datasets = cl.list_datasets()
            print(f"Available datasets: {datasets}")
            
            print("\nWaiting for client connections...")
            print("Press Ctrl+C to stop the server")
            
            # Keep the server running until interrupted
            while True:
                time.sleep(1)
                
        except KeyboardInterrupt:
            print("\nStopping server...")
        finally:
            # Clean up
            cl.stop_flight_server()
            print("Flight server stopped")
            
    elif args.mode == 'client':
        # Client mode: connect to a Flight server and pull data
        print(f"Connecting to server at {args.host}:{args.port}")
        
        # Create a configuration for local mode
        config = CrossLinkConfig(
            db_path="crosslink_client.duckdb",
            debug=True
        )
        
        # Initialize CrossLink
        cl = get_instance(config=config)
        
        try:
            # List datasets on the remote server
            print("\nListing datasets on remote server...")
            datasets = cl.list_remote_datasets(args.host, args.port)
            print(f"Available datasets: {datasets}")
            
            if not datasets:
                print("No datasets available on the server.")
                return 0
                
            # Pull the first dataset
            dataset_id = datasets[0]
            print(f"\nPulling dataset '{dataset_id}' from remote server...")
            table = cl.flight_pull(dataset_id, args.host, args.port)
            
            print("\nPulled table:")
            print(table.to_pandas())
            
            # Create a new table to push back to the server
            new_df = pd.DataFrame({
                'id': [101, 102, 103],
                'name': ['Xavier', 'Yolanda', 'Zach'],
                'value': [99.9, 88.8, 77.7]
            })
            new_table = pa.Table.from_pandas(new_df)
            
            print("\nPushing a new table to the server...")
            print(new_df)
            
            # Push the table to the remote server
            new_dataset_id = cl.flight_push(new_table, args.host, args.port, 
                                         "client_dataset", "Dataset from client")
            print(f"Table pushed with ID: {new_dataset_id}")
            
            # List datasets again to verify the new dataset
            datasets = cl.list_remote_datasets(args.host, args.port)
            print(f"Updated datasets on server: {datasets}")
            
        except Exception as e:
            print(f"Error: {e}")
            return 1

    return 0

if __name__ == "__main__":
    sys.exit(main()) 

File: duckdata/crosslink/cpp/examples/flight_example.cpp
#include <crosslink/crosslink.h>
#include <arrow/api.h>
#include <arrow/table.h>
#include <iostream>
#include <vector>
#include <thread>
#include <chrono>

using namespace crosslink;

int main() {
    try {
        // 1. Initialize Node A (server node)
        std::cout << "Initializing Node A (server)..." << std::endl;
        CrossLinkConfig config_a;
        config_a.set_mode(OperationMode::DISTRIBUTED)
                .set_flight_host("localhost")
                .set_flight_port(8815)
                .set_debug(true);
        
        CrossLink node_a(config_a);
        
        // Start Flight server on Node A
        if (!node_a.start_flight_server()) {
            std::cerr << "Failed to start Flight server on Node A" << std::endl;
            return 1;
        }
        
        int server_port = node_a.flight_server_port();
        std::cout << "Node A Flight server started on port " << server_port << std::endl;
        
        // 2. Create a sample table on Node A
        std::cout << "\nCreating sample table on Node A..." << std::endl;
        auto schema = arrow::schema({arrow::field("id", arrow::int64()), 
                                    arrow::field("name", arrow::utf8())});
        
        arrow::Int64Builder id_builder;
        arrow::StringBuilder name_builder;
        
        id_builder.AppendValues({1, 2, 3, 4, 5});
        name_builder.AppendValues({"Alice", "Bob", "Charlie", "David", "Eve"});
        
        std::shared_ptr<arrow::Array> id_array, name_array;
        id_builder.Finish(&id_array);
        name_builder.Finish(&name_array);
        
        auto table = arrow::Table::Make(schema, {id_array, name_array});
        
        // 3. Push the table to Node A's local storage
        std::string dataset_id = node_a.push(table, "sample_dataset", "Sample dataset for Flight test");
        std::cout << "Table pushed to Node A with ID: " << dataset_id << std::endl;
        
        // 4. Initialize Node B (client node)
        std::cout << "\nInitializing Node B (client)..." << std::endl;
        CrossLinkConfig config_b;
        config_b.set_mode(OperationMode::LOCAL)
                .set_debug(true);
        
        CrossLink node_b(config_b);
        
        // 5. Pull the table from Node A to Node B using Flight
        std::cout << "\nNode B pulling table from Node A via Flight..." << std::endl;
        auto remote_table = node_b.flight_pull(dataset_id, "localhost", server_port);
        
        std::cout << "Table pulled by Node B: " << remote_table->num_rows() << " rows, "
                 << remote_table->num_columns() << " columns" << std::endl;
        
        // 6. List datasets on Node A from Node B
        std::cout << "\nNode B listing datasets on Node A..." << std::endl;
        auto datasets = node_b.list_remote_datasets("localhost", server_port);
        
        std::cout << "Datasets on Node A:" << std::endl;
        for (const auto& id : datasets) {
            std::cout << " - " << id << std::endl;
        }
        
        // 7. Create another table on Node B
        std::cout << "\nCreating another table on Node B..." << std::endl;
        arrow::Int64Builder id_builder2;
        arrow::StringBuilder name_builder2;
        
        id_builder2.AppendValues({101, 102, 103});
        name_builder2.AppendValues({"Xavier", "Yolanda", "Zach"});
        
        std::shared_ptr<arrow::Array> id_array2, name_array2;
        id_builder2.Finish(&id_array2);
        name_builder2.Finish(&name_array2);
        
        auto table2 = arrow::Table::Make(schema, {id_array2, name_array2});
        
        // 8. Push the table from Node B to Node A using Flight
        std::cout << "\nNode B pushing table to Node A via Flight..." << std::endl;
        std::string dataset_id2 = node_b.flight_push(table2, "localhost", server_port, 
                                                 "node_b_dataset", "Dataset from Node B");
        
        std::cout << "Table pushed from Node B to Node A with ID: " << dataset_id2 << std::endl;
        
        // 9. Pull the table back on Node A to confirm
        std::cout << "\nNode A pulling back the table from its own storage..." << std::endl;
        auto confirmed_table = node_a.pull(dataset_id2);
        
        std::cout << "Table confirmed on Node A: " << confirmed_table->num_rows() << " rows, "
                 << confirmed_table->num_columns() << " columns" << std::endl;
        
        // 10. Stop the Flight server
        std::cout << "\nStopping Node A Flight server..." << std::endl;
        node_a.stop_flight_server();
        
        std::cout << "\nFlight example completed successfully!" << std::endl;
        return 0;
        
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
} 